src/java.base/share/classes/jdk/crac/RCULock.java

```diff
@@ -0,0 +1,325 @@
+     *
+     * @apiNote This constructor does not use var-args argument to prevent
+     * accidentally creating a RCULock without declaring any critical methods.
+     *
+     * @param readCriticalMethods List of signatures for methods invoked in the read-critical section.
+     */
+    public RCULock(String[] readCriticalMethods) {
+        initSwitchPoints();
+        methodsCopy = new TreeSet<>(Arrays.asList(readCriticalMethods));
+        //noinspection CapturingCleaner
+        CleanerFactory.cleaner().register(this, this::destroy);
+        Arrays.sort(readCriticalMethods);
+        init(readCriticalMethods);
+    }
+
+    private native void init(String[] readCriticalMethods);
+
+    private native void destroy();
+
+    /**
+     * Inspect the classes for all methods (including private ones) marked with
+     * the {@link RCULock.Critical} annotation and add these to the list
+     * of read-critical methods.
+     *
+     * @param classes Classes to be inspected.
+     */
+    public void amendClasses(Class<?>... classes) {
+        amendCriticalMethods(findAnnotated(classes));
+    }
+
+    /**
+     * Add read-critical methods.
+     *
+     * @param methods List of read-critical methods.
+     *
+     * @see RCULock#RCULock(Iterable)
+     */
+    public void amendCriticalMethods(Iterable<Method> methods) {
+        amendCriticalMethods(StreamSupport.stream(methods.spliterator(), false)
+                .map(RCULock::signature).toArray(String[]::new));
+    }
+
+    /**
+     * Add read-critical methods using signatures.
+     *
+     * @param methods List of method signatures.
+     *
+     * @see RCULock#RCULock(String[])
+     */
+    public void amendCriticalMethods(String... methods) {
+        lock.lock();
+        try {
+            if (synchronize) {
+                throw new IllegalMonitorStateException("Already synchronizing");
+            }
+            methodsCopy.addAll(Arrays.asList(methods));
+            updateMethods(methodsCopy.toArray(new String[0]));
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private native void updateMethods(String[] methods);
+
+    private void initSwitchPoints() {
+        try {
+            MethodType voidType = MethodType.methodType(void.class);
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+            MethodHandle noop = lookup.findSpecial(RCULock.class, "noop", voidType, RCULock.class);
+            MethodHandle readLockImpl = lookup.findSpecial(RCULock.class, "readLockImpl", voidType, RCULock.class);
+            MethodHandle readUnlockImpl = lookup.findSpecial(RCULock.class, "readUnlockImpl", voidType, RCULock.class);
+            lockSwitchPoint = new SwitchPoint();
+            lockImpl = lockSwitchPoint.guardWithTest(noop, readLockImpl);
+            unlockSwitchPoint = new SwitchPoint();
+            unlockImpl = lockSwitchPoint.guardWithTest(noop, readUnlockImpl);
```

Comment:
After spending a lot of time digging through the Hotspot C2 code, I don't think this use of SwitchPoints will be optimized in the way we might want (ie: deopt on SwitchPoint invalidation).

SwitchPoint is implemented as a special MutableCallSite that changes from a known target MH that returns true to a known target MH that returns false.  This is all wrapped into a standard guardWithTest MH (basically a MH "if" statement).

As far as I can tell, to benefit from C2 optimizing the SW, we need to have the MutableCallSite or the GuardWithTest MH rooted in a static final field or in an invokedynamic bytecode. See the code in type.cpp::make_constant_from_field which asserts a Dependency between the CallSite and the target MH for the current method.  Otherwise, we don't have a Dependency to rely on and trigger the eventual deopt.

The above analysis may be wrong so any corrections by C2 experts would be appreciated.

Reply:
Thanks for the insight! So I guess this could be really simplified to a bunch of `if`s - and it would be more clear to the reader, as he wouldn't expect some mysterious magic reason to pick SP. I'll rerun my benchmarks after that and we'll see if there's any difference.

While I an imagine limiting this to a `static` field (coarse-graining this to singleton for the whole VM - it's not too bad anyway since we execute a VM-wide operation), but I can't make this `final` as we need to flip this many times.